File: compiler/src/main/java/edu/cornell/cs/apl/viaduct/security/solver/LabelConstant.java
package edu.cornell.cs.apl.viaduct.security.solver;

import com.google.auto.value.AutoValue;
import edu.cornell.cs.apl.viaduct.algebra.FreeDistributiveLattice;
import edu.cornell.cs.apl.viaduct.algebra.solver.ConstantTerm;
import edu.cornell.cs.apl.viaduct.security.Label;
import edu.cornell.cs.apl.viaduct.security.Principal;
import java.util.Map;

/** Terms representing literal label constants. */
@AutoValue
public abstract class LabelConstant extends AtomicLabelTerm {
  /** Create a term that represents the given value. */
  public static LabelConstant create(Label value) {
    return new AutoValue_LabelConstant(value);
  }

  public abstract Label getValue();

  @Override
  public final Label getValue(Map<LabelVariable, Label> solution) {
    return getValue();
  }

  @Override
  public LabelConstant confidentiality() {
    return create(getValue().confidentiality());
  }

  @Override
  public LabelConstant integrity() {
    return create(getValue().integrity());
  }

  @Override
  public final LabelConstant swap() {
    return create(getValue().swap());
  }

  @Override
  public final LabelConstant join(Label that) {
    return create(getValue().join(that));
  }

  @Override
  final ConstantTerm<FreeDistributiveLattice<Principal>> getConfidentialityComponent() {
    return new ConstantTerm<>(getValue().getConfidentialityComponent());
  }

  @Override
  final ConstantTerm<FreeDistributiveLattice<Principal>> getIntegrityComponent() {
    return new ConstantTerm<>(getValue().getIntegrityComponent());
  }
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/security/solver/LabelConstant.kt
package edu.cornell.cs.apl.viaduct.security.solver

import edu.cornell.cs.apl.viaduct.algebra.FreeDistributiveLattice
import edu.cornell.cs.apl.viaduct.algebra.solver.ConstantTerm
import edu.cornell.cs.apl.viaduct.security.Label
import edu.cornell.cs.apl.viaduct.security.Principal

/** Terms representing literal label constants. */
data class LabelConstant(val value: Label) : AtomicLabelTerm() {
    override fun getValue(solution: ConstraintSolution): Label {
        return value
    }

    override fun confidentiality(): LabelConstant {
        return LabelConstant(value.confidentiality())
    }

    override fun integrity(): LabelConstant {
        return LabelConstant(value.integrity())
    }

    override fun swap(): LabelConstant {
        return LabelConstant(value.swap())
    }

    override fun join(that: Label): LabelConstant {
        return LabelConstant(value.join(that))
    }

    override val confidentialityComponent: ConstantTerm<FreeDistributiveLattice<Principal>>
        get() = ConstantTerm(value.confidentialityComponent)

    override val integrityComponent: ConstantTerm<FreeDistributiveLattice<Principal>>
        get() = ConstantTerm(value.integrityComponent)
}
